usual/tls/tls.c

```diff
@@ -742,4 +742,84 @@ tls_close(struct tls *ctx)
 	return (rv);
 }
 
+static bool
+tls_config_keypair_equal(struct tls_keypair *left_tls_keypair, struct tls_keypair *right_tls_keypair)
+{
+  if (!strings_equal(left_tls_keypair->cert_file, right_tls_keypair->cert_file))
+    return false;
+  if (left_tls_keypair->cert_mem && right_tls_keypair->cert_mem &&
+      !(memcmp(left_tls_keypair->cert_mem, right_tls_keypair->cert_mem, left_tls_keypair->cert_len) == 0))
+    return false;
+  if (!strings_equal(left_tls_keypair->key_file, right_tls_keypair->key_file))
+    return false;
+  if (left_tls_keypair->key_mem && right_tls_keypair->key_mem &&
+      !(memcmp(left_tls_keypair->key_mem, right_tls_keypair->key_mem, left_tls_keypair->key_len) != 0))
+    return false;
+  return true;
+}
+
+static bool
+tls_config_keypairs_equal(struct tls_keypair *tls_keypair_left, struct tls_keypair *tls_keypair_right)
+{
+	bool keypair_unchanged;
+	struct tls_keypair *tls_kp_lt, *next_tls_kp_lt, *tls_kp_rt, *next_tls_kp_rt;
+
+	keypair_unchanged = tls_config_keypair_equal(tls_keypair_left, tls_keypair_right);
+
+	if (keypair_unchanged == false) {
+		return false;
+	}
+
+	for (tls_kp_lt = tls_keypair_left; tls_kp_lt != NULL; tls_kp_lt = next_tls_kp_lt) {
+		next_tls_kp_lt = tls_kp_lt->next;
+		for (tls_kp_rt = tls_keypair_right; tls_kp_rt != NULL; tls_kp_rt = next_tls_kp_rt) {
+			next_tls_kp_rt = tls_kp_rt->next;
+
+			keypair_unchanged = tls_config_keypair_equal(tls_kp_lt, tls_kp_rt);
+		}
+	}
+
+	return keypair_unchanged;
+}
+
+bool
+tls_configs_equal(struct tls_config *tls_config_left, struct tls_config *tls_config_right)
+{
+  if (!(strings_equal(tls_config_left->ca_file, tls_config_right->ca_file)))
+    return false;
+  if (!(strings_equal(tls_config_left->ca_path, tls_config_right->ca_path)))
+    return false;
+  if (tls_config_left->ca_mem &&  tls_config_right->ca_mem &&
+      !(memcmp(tls_config_left->ca_mem, tls_config_right->ca_mem, tls_config_left->ca_len) == 0))
+    return false;
+  if (!(strings_equal(tls_config_left->ciphers, tls_config_right->ciphers)))
+    return false;
+  if (!(tls_config_left->ciphers_server == tls_config_right->ciphers_server))
```

Comment:
I do appreciate the consistency of having all these statements contain a `!` at the start. But afaik it's used nowhere else in the codebase. So I prefer to stay consistent with the style we already have, which would be no parenthesis around `xxx_equal` functions and using `x != y` instead of `!(x == y)`. So like this:
```suggestion
  if (!strings_equal(tls_config_left->ciphers, tls_config_right->ciphers))
    return false;
  if (tls_config_left->ciphers_server != tls_config_right->ciphers_server)
```

Reply:
Thanks 16ff462

Reply:
I would suggest short variable names. I prefer short names when the number of variables is small, hence, it is easier to figure out what it is for.

```
-  if (!(strings_equal(tls_config_left->ca_file, tls_config_right->ca_file)))
-    return false;
+   if (!strcmpeq(tc1->ca_file, tc2->ca_file))
+       return false;
```

Reply:
Ah, I am personally the polar opposite on variable naming but it's more important to me to conform to a code-bases' conventions and team's expectations so I'm quite happy to make this change.

Reply:
2473167